### PR TITLE
feat: add timestamp for DAP trace and formate output using `JSON.stringify(message, null, 4)`

### PR DESCRIPTION
**Pull request by DanboDuan**
#### ADD timestamp for DAP trace output
DAP trace output does not contain timestamp. it looks like this

```
fabf8f58 theia <- adapter: {"body":{"reason":"exited","threadId":2305159},"event":"thread","seq":0,"type":"event"}
fabf8f58 theia <- adapter: {"body":{"reason":"exited","threadId":2305187},"event":"thread","seq":0,"type":"event"}
fabf8f58 theia <- adapter: {"body":{"reason":"exited","threadId":2305194},"event":"thread","seq":0,"type":"event"}
fabf8f58 theia <- adapter: {"body":{"reason":"exited","threadId":2305286},"event":"thread","seq":0,"type":"event"}
fabf8f58 theia <- adapter: {"body":{"reason":"exited","threadId":2305343},"event":"thread","seq":0,"type":"event"}
fabf8f58 theia <- adapter: {"body":{"reason":"exited","threadId":2305191},"event":"thread","seq":0,"type":"event"}
fabf8f58 theia <- adapter: {"body":{"reason":"exited","threadId":2305281},"event":"thread","seq":0,"type":"event"}
fabf8f58 theia -> adapter: {"seq":18,"type":"request","command":"threads","arguments":{}}
```
It's not easy to get the performance with those trace output. It's better to output a timestamp so you will be easy to get a context.
#### How to test

check the dap trace output is now with timestamp like this

```
fabf8f58 theia <- adapter: {"body":{"reason":"exited","threadId":2305159},"event":"thread","seq":0,"type":"event"}
```

after the changes submit, check the dap trace output is now with timestamp like this

```
fabf8f58 2021-11-27T02:40:06.428Z theia <- adapter: {"body":{"reason":"exited","threadId":2305159},"event":"thread","seq":0,"type":"event"}
```
#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/eclipse-theia/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with the review guidelines

### CODE REVIEW

File: packages/debug/src/browser/debug-session-connection.ts [hunk 1]
```diff
@@ -243,14 +243,22 @@ export class DebugSessionConnection implements Disposable {
         const connection = await this.connectionPromise;
         const messageStr = JSON.stringify(message);
         if (this.traceOutputChannel) {
-            this.traceOutputChannel.appendLine(`${this.sessionId.substring(0, 8)} theia -> adapter: ${messageStr}`);
+            const now = new Date();
+            /// offset in milliseconds
+            const tzoffset = now.getTimezoneOffset() * 60 * 1000;
+            const dateStr = new Date(now.getTime() - tzoffset).toISOString();
```
Comment (msujew) [inline packages/debug/src/browser/debug-session-connection.ts:249]: Suggestion: As far as I know, `toISOString` will always convert the input time into UTC. So we can simplify this by using the following code. The same applies to the other change.
```suggestion
            const dateStr = new Date().toISOString();
```

**Review by msujew — COMMENTED** [review]
Hi @DanboDuan, thank you for your first contribution.

the improvement to logging looks good. I have a suggestion to simplify the change.

File: packages/debug/src/browser/debug-session-connection.ts (hunk 1 above)
Comment (DanboDuan) [reply packages/debug/src/browser/debug-session-connection.ts:249]: minus the timezone offset for the output to be the same as local time. If not, the time string should be `2021-11-27T02:40:06.428Z` for a UTC+9 which should be `2021-11-27 11:40:06.428` as local. It is good, but it looks strange. So I make a new date minus the timezone offset to create a data string

Comment (DanboDuan) [reply packages/debug/src/browser/debug-session-connection.ts]: I update to use `toLocaleString`

File: packages/debug/src/browser/debug-session-connection.ts [hunk 2]
```diff
@@ -243,16 +243,22 @@ export class DebugSessionConnection implements Disposable {
         const connection = await this.connectionPromise;
         const messageStr = JSON.stringify(message);
         if (this.traceOutputChannel) {
-            this.traceOutputChannel.appendLine(`${this.sessionId.substring(0, 8)} theia -> adapter: ${messageStr}`);
+            // keep getMilliseconds
+            const now = new Date();
+            const dateStr = `${now.toLocaleString(undefined, { hour12: false })}.${now.getMilliseconds()}`;
+            this.traceOutputChannel.appendLine(`${this.sessionId.substring(0, 8)} ${dateStr} theia -> adapter: ${JSON.stringify(message, null, 4)}`);
```
Comment (vince-fugnitto) [inline packages/debug/src/browser/debug-session-connection.ts:249]: @DanboDuan if you don't need `null` explicitly, please use `undefined` instead, it fails the build unless the line is suppressed:

```suggestion
            this.traceOutputChannel.appendLine(`${this.sessionId.substring(0, 8)} ${dateStr} theia -> adapter: ${JSON.stringify(message, undefined, 4)}`);
```

The same is similar for `handleMessage`.

Comment (DanboDuan) [reply packages/debug/src/browser/debug-session-connection.ts]: Thx, done.

File: packages/debug/src/browser/debug-session-connection.ts [hunk 3]
```diff
@@ -243,16 +243,22 @@ export class DebugSessionConnection implements Disposable {
         const connection = await this.connectionPromise;
         const messageStr = JSON.stringify(message);
         if (this.traceOutputChannel) {
-            this.traceOutputChannel.appendLine(`${this.sessionId.substring(0, 8)} theia -> adapter: ${messageStr}`);
+            // keep getMilliseconds
```
Comment (vince-fugnitto) [inline packages/debug/src/browser/debug-session-connection.ts:246]: I believe we can remove the comment, it does not add value (the same goes for `handleMessage`).

**Review by vince-fugnitto — APPROVED** [review]
I confirmed that the changes work well, and the timestamp is now added to the traces of the debug adapter. Example:

```
ab3a895f 12/1/2021, 08:50:30.582 theia -> adapter: {
    "seq": 34,
    "type": "request",
    "command": "scopes",
    "arguments": {
        "frameId": 34
    }
}
```

File: packages/debug/src/browser/debug-session-connection.ts (hunk 3 above)
Comment (DanboDuan) [reply packages/debug/src/browser/debug-session-connection.ts]: ok

**Comment by vince-fugnitto** [comment]
@msujew any objections? :)

**Review by msujew — APPROVED** [review]
@vince-fugnitto None, looks good to me 👍

**Pull request merged**